### PR TITLE
Config Additions: session_recording, mask_all_element_attributes, mask_all_text

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Changes

So I was operating on a few hours of sleep today and put through #207 which, while preventing rrweb from capturing input values as stated, it doesn't prevent PostHog from doing so, because..... we already don't. 🤦  I confused our approach with rrweb's and went ahead and censored element attributes instead.

Nevertheless, this PR now adds multiple config options:

- `session_recording`: an object containing a subset of `rrweb` options that we can expose to our users
- `mask_all_text`: Prevents us from capturing `textContent` as per @lharress's request (and this makes a lot of sense)
- `mask_all_element_attributes`: prevents autocapture from capturing **any** element attributes (apparently you'd need this too @lharress?)

## Checklist
- [x] Tests for new code (if applicable)
- [x] TypeScript definitions (module.d.ts) updated and in sync with library exports (if applicable)

### CODE REVIEW

**Comment by lharress** [comment]
Thanks, I think you missed one?

https://github.com/PostHog/posthog-js/blob/7faa07b53b12c0eda75085a2aeb6991d8c197934/src/autocapture.js#L35

The second kinda works for us (insert sledgehammer scalpel analogy here) in that we have some custom attributes that don't fall within the `input` sensitivity check. The 3 options (exclude attributes, textContent and the built in `input` conservativeness) will definitely help us. Thanks!

**Comment by yakkomajuri** [comment]
Fixed, thanks a lot for keeping an eye on this!

@posthog-bot please add @lharress for ideas and test

**Review by macobo — COMMENTED** [review]
We currently lack any E2E testing for autocapture functionality (and this would have introduced a bug there) - how do you feel about extending the testcafe test for this?

File: src/autocapture.js [hunk 1]
```diff
@@ -175,12 +177,18 @@ var autocapture = {
...
+                            instance.get_config('mask_all_element_attributes'),
+                            instance.get_config('mask_all_text')
+                        )
+                    )
                 },
                 this
             )
 
-            elementsJson[0]['$el_text'] = getSafeText(target)
+            elementsJson[0]['$el_text'] = instance.get_config('mask_all_text') ? '' : getSafeText(target)
```
[the remark this answers is not in the document]
Comment (yakkomajuri) [reply src/autocapture.js:191]: makes sense

**Comment by yakkomajuri** [comment]
Thanks! I have no time today but can hopefully find time to improve the tests tomorrow!

**Comment by yakkomajuri** [comment]
I've addressed everything except testcafe tests - can get to it on this PR or a later one. 

cc @macobo

**Comment by macobo** [comment]
Might be worth getting in here, but would also be OK with merging as-is if you're confident in it. :)

**Comment by yakkomajuri** [comment]
Comment came before I had access to browserstack - I now do so I'll play around with it

File: testcafe/e2e.spec.js [hunk 2]
```diff
@@ -34,3 +34,46 @@ test('Captured events are accessible via /api/event', async (t) => {
     await t.expect(results.filter(({ event }) => event === 'custom-event').length).gte(1)
     await t.expect(results.filter(({ event }) => event === '$pageview').length).gte(1)
 })
+
+test('Config option mask_all_text does not set $el_text for autocaptured events', async (t) => {
```
Comment (macobo) [inline testcafe/e2e.spec.js:38]: We're testing a negative of something that's not tested at all here.

To bring most value to the E2E suite, how about:
- One test (or extending the existing test) confirming that autocapture works as expected (e.g. captures attributes and text and whatelse)
- One test for _some_ of the negative conditions.

Generally it's really hard to test all cases E2E and these tests can be brittle - better to test these via unit tests. However autocapture is a cruicial area which we hadn't changed until now and had no e2e test coverage.

File: cypress/integration/capture.spec.js [hunk 3]
```diff
@@ -55,7 +55,7 @@ describe('Event capture', () => {
 
         cy.get('body').click(100, 100).click(98, 102).click(101, 103)
 
-        cy.phCaptures().should('deep.equal', ['$pageview', '$rageclick'])
```
Comment (macobo) [inline cypress/integration/capture.spec.js:58]: Q: Why did this break?

Comment (yakkomajuri) [reply cypress/integration/capture.spec.js:58]: You're too fast! I've fixed it now. Essentially, by adding more elements to the playground html, we're rendering this:

Screenshot 2021-04-06 at 13 54 55

So clicks a few pixels off were also clicking other elements, and hence that array had more events than just those two.

File: testcafe/e2e.spec.js (hunk 2 above)
Comment (yakkomajuri) [reply testcafe/e2e.spec.js:38]: I can do that, yeah

**Comment by yakkomajuri** [comment]
Definitely happy to update/change this again, but I've now structured E2E tests in 3 parts, to "set the tone" for future tests:

- Custom events work
- Autocapture works (captures events and the right props)
- Config options correctly affect capture behavior

**Review by macobo — APPROVED** [review]

**Pull request merged**